File: pipeline/nfl_pipeline/contributions.py
"""Port of teams_tab.py:compute_week_contributions + feature-set helpers.

Runs for every (team, season, week, grade_type) so the frontend never needs
to fit scalers at runtime.
"""
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

DESC_COLS = {'team', 'season', 'week', 'season_type', 'opponent_team', 'gameday', 'game_id', 'game_type'}


def _off_cols(df: pd.DataFrame):
    drop_cols = DESC_COLS | {'total_tds', 'points_margin', 'fantasy_points', 'fantasy_points_ppr', 'win', 'points'}
    numeric = df.select_dtypes(include='number').columns
    return [c for c in numeric if c not in drop_cols and not c.endswith("_allowed")]


def _def_cols(df: pd.DataFrame):
    numeric = df.select_dtypes(include='number').columns
    return [c for c in numeric if c.endswith("_allowed") and c != 'points_allowed']


def _overall_cols(df: pd.DataFrame):
    drop_cols = DESC_COLS | {'total_tds', 'fantasy_points', 'fantasy_points_ppr', 'points',
                             'points_allowed', 'epa_diff', 'win', 'points_margin'}
    numeric = df.select_dtypes(include='number').columns
    return [c for c in numeric if c not in drop_cols]


def _select_cols(df: pd.DataFrame, grade_type: str):
    if grade_type == "Offensive Grade":
        return _off_cols(df), "Offensive Importance", "offense"
    if grade_type == "Defensive Grade":
        return _def_cols(df), "Defensive Importance", "defense"
    return _overall_cols(df), "Overall Importance", "overall"


def contribution_params(team_week_df: pd.DataFrame, weights_df: pd.DataFrame) -> dict:
    """Export the parameters the frontend needs to reproduce
    compute_week_contributions exactly, instead of materializing ~1.6M rows.

    For each grade type: the feature list (order preserved), MinMax data_min /
    data_max fitted on ALL rows (identical to teams_tab.py fitting the scaler
    on the whole league frame), the importance weight, and the mode.
    Frontend math: norm = clip-free (raw - min)/(max - min) [sklearn transform,
    no clipping]; signed = w*(1-norm) for defense else w*norm.
    """
    league = team_week_df.reset_index(drop=True)
    out = {}
    for grade_type in ["Offensive Grade", "Defensive Grade", "Overall Grade"]:
        feat_cols, weight_col, mode = _select_cols(league, grade_type)
        if not feat_cols or weight_col not in weights_df.columns:
            continue
        X_raw = league[feat_cols].apply(pd.to_numeric, errors="coerce").fillna(0.0)
        scaler = MinMaxScaler().fit(X_raw)

        w = (weights_df[['Feature', weight_col]]
             .dropna(subset=['Feature'])
             .set_index('Feature')
             .reindex(feat_cols)
             .fillna(0.0)[weight_col])

        out[grade_type] = {
            "mode": mode,
            "features": feat_cols,
            "data_min": [float(v) for v in scaler.data_min_],
            "data_max": [float(v) for v in scaler.data_max_],
            "importance": [float(v) for v in w.values],
        }
    return out

File: app/src/lib/logic/contributions.ts
// Port of teams_tab.py:compute_week_contributions using the pipeline's
// precomputed MinMax params + importances (contrib_params.json).
import type { Row, ContribParams } from "../data/loader";

export type GradeType = "Overall Grade" | "Offensive Grade" | "Defensive Grade";

export interface ContribRow {
  feature: string;
  importance: number;
  raw: number;
  norm: number;
  contribution: number; // |signed|
  signed: number;
}

/** Contributions for one team-week row, all features, sorted by |contribution| desc. */
export function weekContributions(twRow: Row, params: ContribParams[string]): ContribRow[] {
  const out: ContribRow[] = params.features.map((f, i) => {
    const raw = twRow[f] == null ? 0 : Number(twRow[f]) || 0; // pandas fillna(0)
    const range = params.data_max[i] - params.data_min[i];
    const norm = (raw - params.data_min[i]) / (range || 1); // sklearn zero-range guard
    const w = params.importance[i];
    const signed = w * (params.mode === "defense" ? 1 - norm : norm);
    return { feature: f, importance: w, raw, norm, contribution: Math.abs(signed), signed };
  });
  return out.sort((a, b) => b.contribution - a.contribution);
}

/** numpy-style linear-interpolation percentile (q in [0,100]) on a sorted copy. */
export function percentile(values: number[], q: number): number {
  const s = [...values].sort((a, b) => a - b);
  if (!s.length) return NaN;
  const idx = (q / 100) * (s.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  return s[lo] + (s[hi] - s[lo]) * (idx - lo);
}

export function sampleStd(values: number[]): number {
  const n = values.length;
  if (n < 2) return 0;
  const mean = values.reduce((a, b) => a + b, 0) / n;
  return Math.sqrt(values.reduce((s, v) => s + (v - mean) ** 2, 0) / (n - 1));
}
